google/cloud/internal/async_streaming_write_rpc_timeout_test.cc

```diff
@@ -0,0 +1,331 @@
+  auto constexpr kWriteTimeout = std::chrono::milliseconds(2345);
+
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  EXPECT_CALL(*mock_cq,
+              MakeRelativeTimer(std::chrono::nanoseconds(kStartTimeout)))
+      .WillOnce([&](auto) {
+        return sequencer.PushBack("MakeRelativeTimer")
+            .then(
+                [](auto f) -> StatusOr<std::chrono::system_clock::time_point> {
+                  if (!f.get()) return CancelledTimer();
+                  return make_status_or(std::chrono::system_clock::now());
+                });
+      });
+
+  CompletionQueue cq(mock_cq);
+
+  TestedStream uut(cq, kStartTimeout, kWriteTimeout, std::move(mock));
+  auto result = uut.WritesDone();
+  auto timer = sequencer.PopFrontWithName();
+  auto done = sequencer.PopFrontWithName();
+  ASSERT_THAT(timer.second, "MakeRelativeTimer");
+  ASSERT_THAT(done.second, "WritesDone");
+  done.first.set_value(true);
+  timer.first.set_value(false);
+  EXPECT_TRUE(result.get());
+}
+
+TEST(AsyncStreamingWriteRpcTimeout, WritesDoneTimeout) {
+  AsyncSequencer<bool> sequencer;
+  auto mock = std::make_unique<MockStream>();
+  EXPECT_CALL(*mock, WritesDone).WillOnce([&] {
+    return sequencer.PushBack("WritesDone");
+  });
+  EXPECT_CALL(*mock, Cancel).Times(1);
+
+  auto constexpr kStartTimeout = std::chrono::milliseconds(2345);
+  auto constexpr kWriteTimeout = std::chrono::milliseconds(2345);
+
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  EXPECT_CALL(*mock_cq,
+              MakeRelativeTimer(std::chrono::nanoseconds(kStartTimeout)))
+      .WillOnce([&](auto) {
+        return sequencer.PushBack("MakeRelativeTimer")
+            .then(
+                [](auto f) -> StatusOr<std::chrono::system_clock::time_point> {
+                  if (!f.get()) return CancelledTimer();
+                  return make_status_or(std::chrono::system_clock::now());
+                });
+      });
+
+  CompletionQueue cq(mock_cq);
+
+  TestedStream uut(cq, kStartTimeout, kWriteTimeout, std::move(mock));
+  auto result = uut.WritesDone();
+  auto timer = sequencer.PopFrontWithName();
+  auto done = sequencer.PopFrontWithName();
+  ASSERT_THAT(timer.second, "MakeRelativeTimer");
+  ASSERT_THAT(done.second, "WritesDone");
+  timer.first.set_value(true);
+  done.first.set_value(true);
+  EXPECT_FALSE(result.get());
+}
+
+TEST(AsyncStreamingWriteRpcTimeout, Finish) {
+  AsyncSequencer<bool> sequencer;
+  auto mock = std::make_unique<MockStream>();
+  EXPECT_CALL(*mock, Finish).WillOnce([&] {
+    return sequencer.PushBack("Finish").then(
+        [](auto f) -> StatusOr<std::string> {
+          if (!f.get())
+            return internal::DataLossError("uh oh", GCP_ERROR_INFO());
+          return std::string{"42"};
+        });
+  });
+  EXPECT_CALL(*mock, Cancel).Times(0);
+
+  auto constexpr kStartTimeout = std::chrono::milliseconds(2345);
+  auto constexpr kReadTimeout = std::chrono::milliseconds(2345);
```

Comment:
nit: s/Read/Write/

Reply:
optional: Factor these out of the individual tests, and save some lines?

Reply:
> nit: s/Read/Write/

Done.

> optional: Factor these out of the individual tests, and save some lines?

Done. Though sometimes it is easier to grok a test if you don't need to understand all the refactored code in the test file.